libnopegl/src/backends/gl/pipeline_gl.c

```diff
@@ -744,23 +744,34 @@ static void get_scissor(struct pipeline *s, int32_t *scissor)
     }
 }
 
-void ngli_pipeline_gl_draw(struct pipeline *s, int nb_vertices, int nb_instances)
+static void set_graphics_state(struct pipeline *s)
 {
-    struct pipeline_gl *s_priv = (struct pipeline_gl *)s;
     struct gpu_ctx *gpu_ctx = s->gpu_ctx;
     struct gpu_ctx_gl *gpu_ctx_gl = (struct gpu_ctx_gl *)gpu_ctx;
     struct glcontext *gl = gpu_ctx_gl->glcontext;
     struct glstate *glstate = &gpu_ctx_gl->glstate;
     struct pipeline_graphics *graphics = &s->graphics;
-    struct program_gl *program_gl = (struct program_gl *)s->program;
-
-    s_priv->insert_memory_barriers(s);
 
     ngli_glstate_update(gl, glstate, &graphics->state);
     int32_t scissor[4];
     get_scissor(s, scissor);
     ngli_glstate_update_viewport(gl, glstate, gpu_ctx_gl->viewport);
     ngli_glstate_update_scissor(gl, glstate, scissor);
+}
+
+void ngli_pipeline_gl_draw(struct pipeline *s, int nb_vertices, int nb_instances)
+{
+    struct pipeline_gl *s_priv = (struct pipeline_gl *)s;
+    struct gpu_ctx *gpu_ctx = s->gpu_ctx;
+    struct gpu_ctx_gl *gpu_ctx_gl = (struct gpu_ctx_gl *)gpu_ctx;
+    struct glcontext *gl = gpu_ctx_gl->glcontext;
+    struct glstate *glstate = &gpu_ctx_gl->glstate;
+    struct pipeline_graphics *graphics = &s->graphics;
+    struct program_gl *program_gl = (struct program_gl *)s->program;
+
+    s_priv->insert_memory_barriers(s);
+
+    set_graphics_state(s);
```

Comment:
any reason to refactor only the graphic state? The use program and set uniform/buffers/textures as well as bindings seem common as well.

Reply:
Will be done in a separate commit for buffer/texture/attribute but not for uniforms since this part is removed later on. Also, i want to keep the call to use_program() explicit.